### ISSUE TITLE
[bug] conan lock create uses the timestamps of the newest revision, not the locked revision

### ISSUE THREAD

**Issue by sfackler** [conan-io/conan#14108]
### Environment details

* Conan version: 2.0.6
* Python version: 3.10.12

### Steps to reproduce

1. Create a conanfile requiring an old revision of a recipe:
```
[requires]
zlib/1.2.13#13c96f538b52e1600c40b88994de240f
```
2. Run `conan lock create .`
3. Note the timestamp used for the zlib entry in `conan.lock`
4. Run `conan list --remote conancenter --format json "zlib/1/2.13#*"`
5. Note that the timestamp in the lockfile is actually the timestamp of revision e377bee636333ae348d51ca90874e353, not the locked revision 13c96f538b52e1600c40b88994de240f

I'm not 100% sure if this is intentional or not, but it is a bit annoying since it means that re-generation of an existing lockfile without `--update` passed is nondeterministic depending on if a newer version has been uploaded and if the lock conan cache already has the recipe revision.

### Logs

```
$ cat conanfile.txt
[requires]
zlib/1.2.13#13c96f538b52e1600c40b88994de240f
$ conan lock create .

======== Computing dependency graph ========
zlib/1.2.13: Not found in local cache, looking in remotes...
zlib/1.2.13: Checking remote: conancenter
zlib/1.2.13: Downloaded recipe revision 13c96f538b52e1600c40b88994de240f
Graph root
    conanfile.txt: /Users/sfackler/test/conanfile.txt
Requirements
    zlib/1.2.13#13c96f538b52e1600c40b88994de240f - Downloaded (conancenter)

======== Computing necessary packages ========
Requirements
    zlib/1.2.13#13c96f538b52e1600c40b88994de240f:76f7d863f21b130b4e6527af3b1d430f7f8edbea - Missing
Generated lockfile: /Users/sfackler/test/conan.lock
$ cat conan.lock
{
    "version": "0.5",
    "requires": [
        "zlib/1.2.13#13c96f538b52e1600c40b88994de240f%1682597484.674"
    ],
    "build_requires": [],
    "python_requires": []
}
$ conan list --remote conancenter  "zlib/1.2.13#*" --format json
{
    "conancenter": {
        "zlib/1.2.13": {
            "revisions": {
                "40e9a7e11e440715c0f0f3e96448342b": {
                    "timestamp": 1665749201.071
                },
                "647c91ed13c0a6c9ea9add6ca968ea93": {
                    "timestamp": 1666171150.173
                },
                "13c96f538b52e1600c40b88994de240f": {
                    "timestamp": 1667396813.733
                },
                "e377bee636333ae348d51ca90874e353": {
                    "timestamp": 1682597484.674
                }
            }
        }
    }
}
```

**Comment by sfackler** [comment]
This also appears to break `conan lock create --update`:

```
$ cat conanfile.txt
[requires]
zlib/1.2.13
$ rm conan.lock
$ conan lock create --update .

======== Computing dependency graph ========
zlib/1.2.13: Checking remote: conancenter
Graph root
    conanfile.txt: /Users/sfackler/test/conanfile.txt
Requirements
    zlib/1.2.13#13c96f538b52e1600c40b88994de240f - Newer

======== Computing necessary packages ========
Requirements
    zlib/1.2.13#13c96f538b52e1600c40b88994de240f:76f7d863f21b130b4e6527af3b1d430f7f8edbea - Missing
Generated lockfile: /Users/sfackler/test/conan.lock
$ cat conan.lock
{
    "version": "0.5",
    "requires": [
        "zlib/1.2.13#13c96f538b52e1600c40b88994de240f%1682597484.674"
    ],
    "build_requires": [],
    "python_requires": []
}
```

I would expect that to advance to revision e377bee636333ae348d51ca90874e353, but it doesn't, I guess since the local package cache sees 13c96f538b52e1600c40b88994de240f as having the same timestamp?

**Comment by memsharded** [comment]
Hi @sfackler 

Thanks for reporting. This seems a potential bug, lets check it.

A couple of clarifications in the meantime:

> I'm not 100% sure if this is intentional or not, but it is a bit annoying since it means that re-generation of an existing lockfile 

Yes, this is intentional. The timestamps of revisions need to be "local" to the current location, so the timestamp of the package in the cache is not the same as the timestamp of the package in the server. This is very necessary for stability, otherwise the resulting behavior that developers get is very weird and confusing.

> --update passed is nondeterministic depending on if a newer version has been uploaded and if the lock conan cache already has the recipe revision.

It shouldn't be. If you defined the ``recipe revision`` in your ``conanfile.txt``, then the resulting lockfile is guaranteed to be that exact revision (if it is not, it would be a bug), because, in other words, the dependency is already locked, so no much more to lock. 

What is really surprising here is that the timestamp exactly matches the one from the server latest timestamp, that is unexpected and sounds like a bug, lets investigate it.

In any case, please note that a lockfile is nothing more than a snapshot of the versions and revisions of a given dependency graph. If the graph is resolved in the cache, it will lock to the latest versions and revisions found in the cache. If the graph is resolved against a server (because the cache is empty or because ``--update``) it will lock to the versions and revisions found in the server. What I mean is that it is expected that the timestamp on the lockfile can be different than the timestamp of the server, because it should be referring to the local timestamp on the cache.

So lets check that potential timestamp bug, I'll keep you posted, thanks again!

**Comment by sfackler** [comment]
> It shouldn't be. If you defined the `recipe revision` in your `conanfile.txt`, then the resulting lockfile is guaranteed to be that exact revision (if it is not, it would be a bug), because, in other words, the dependency is already locked, so no much more to lock.

Sorry yeah, there I was thinking of the case where you don't specify the recipe revision in the conanfile. I just did that in the example above to more easily generate a lockfile for an older revision.

**Comment by memsharded** [comment]
Update:

It seems that I recalled some behavior incorrectly, and this is what it is said in the tests:

>         # - When a revision is installed from a remote it takes the date from the remote, not
        # updating the date to the current time

And I might have identified the source of the bug in ``RemoteManager`` code:
```python
 zipped_files = self._call_remote(remote, "get_recipe", ref, download_export, metadata,
                                             only_metadata=False)
            # TODO: Optimize this call, it is slow to always query all revisions
            remote_refs = self._call_remote(remote, "get_recipe_revisions_references", ref)
            ref_time = remote_refs[0].timestamp
            ref.timestamp = ref_time
```
That seems to be getting the timestamp of the latest, even if the revision is not the latest. Will try to do a fix for it asap

**Comment by sfackler** [comment]
Nice, thanks!

**Comment by memsharded** [comment]
Submitted a fix in https://github.com/conan-io/conan/pull/14110, for next 2.0.7 (if you know how to run Conan from source from my branch, running it from there, would be useful feedback too. But don't worry if not).

Thanks again for reporting this.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/14110 merged, it will be in next 2.0.7
